python/sqlflow_playground/playground_server_design.md

```diff
@@ -0,0 +1,142 @@
+maintainer can use below command:
+```bash
+    mkdir $HOME/workspace
+    cd $HOME/workspace
+    pip install sqlflow_playground
+    mkdir key_store
+    gen_cert.sh server
+    sqlflow_playground --port=50052 \
+      --ca_crt=key_store/ca/ca.crt \
+      --server_key=key_store/server/server.key \
+      --server_crt=key_store/server/server.crt
+```
+In the above commands, we first installed the sqlflow playground package
+which carries the main cluster operation logic.  Then, we use the key
+tool to generate a server certification file (Of course, it's not necessary
+if you have your own certification files) which enables us to provide
+`https` service.  Finally, we start the `REST API` service at port 50052.
+
+Our playground service uses bi-directional validation.  So, the maintainer
+needs to generate a certification file for a trusted user. Use below command and
+send the generated `.crt` and `.key` file together with the `ca.crt` to
+the user.
+
+```bash
+    gen_cert.sh some_client
+```
+
+### For The User
+
+To use this service, the user should get authorized from the playground's maintainer.
+In detail, user should get `ca.crt`, `client.key` and the `client.crt` file from
+the maintainer and keep them in some very-safe place. Also, the user should ask
+the maintainer for the sqlflow server address and the sqlflow playground server
+address. Then, the user will install Jupyter Notebook and the SQLFlow plugin package
+and do some configuration. Finally, the user can experience SQLFlow in his Jupyter 
+Notebook.
+
+```bash
+    pip3 install notebook sqlflow==0.15.0
+
+    cat >$HOME/.sqlflow_playground.env <<EOF
+SQLFLOW_SERVER="{sqlflow server address}"
+SQLFLOW_PLAYGROUND_USER_ID_ENV=SQLFLOW_USER_ID
+SQLFLOW_USER_ID="{your name}"
+SQLFLOW_PLAYGROUND_SERVRE="{sqlflow playground server address}"
+SQLFLOW_PLAYGROUND_SERVER_CA="{path to your ca.crt file}"
+SQLFLOW_PLAYGROUND_CLIENT_KEY="{path to your client.key file}"
+SQLFLOW_PLAYGROUND_CLIENT_CERT="{path to your client.crt file}"
+EOF
+
+    export SQLFLOW_JUPYTER_ENV_PATH="$HOME/.sqlflow_playground.env"
+    # start the notebook and try use %%sqlflow magic command
+    jupyter notebook
+```
+
+## Implementation
+
+We use [tornado](https://www.tornadoweb.org/) as the web framework which provides
+a very good request dispatching mechanism. By the way, this framework is also
+adopted by Jupyter Notebook. The request processing is split into two steps:
+
+1. Register a request handler
+
+    ```python
+    tornado.web.Application([(r"/", MainHandler)])
+    ```
+1. Implement the handler as a class, the method name `get` imply
+    it accepts `GET` requests.
+
+    ```python
+    class MainHandler(RequestHandler):
+        def get(self):
+           self.write("hello SQLFlow!") 
+    ```
+In addition, We add a k8s manipulate class, which can create resource in the
+cluster. It's now implemented in a brutal way (use kubectl). We may refine it
+by using k8s's API.
+
+## TODO
```

Comment:
Maybe we should remove this **TODO** section, and forces on the current design of the playground server.
If the following features are needed in the future, then we can update this design doc.